rpc/src/rpc/native/block.rs

```diff
@@ -0,0 +1,63 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use ethers::prelude::*;
+use ethers::providers::{Http, Provider};
+use ethers::types::{H160, H256};
+use futures::stream::{self, TryStreamExt};
+use tokio::sync::Mutex;
+use trace_decoder::trace_protocol::{AtomicUnitInfo, BlockTrace, TriePreImage, TxnInfo};
+
+/// Processes the block with the given block number and returns the block trace.
+pub async fn process_block_trace(
+    provider: Arc<Provider<Http>>,
+    block_number: u64,
+) -> Result<BlockTrace> {
+    let block = provider
+        .get_block(block_number)
+        .await?
+        .ok_or_else(|| anyhow!("Block not found. Block number: {}", block_number))?;
+
+    let mut accounts_state = HashMap::<H160, HashSet<H256>>::new();
+
+    block.withdrawals.as_ref().map(|w| {
+        w.iter().for_each(|w| {
+            accounts_state.insert(w.address, Default::default());
+        })
+    });
+    accounts_state.insert(
+        block
+            .author
+            .ok_or_else(|| anyhow!("Block author not found"))?,
+        Default::default(),
+    );
+
+    let accounts_state = Arc::new(Mutex::new(accounts_state));
+    let code_db = Arc::new(Mutex::new(HashMap::<H256, Vec<u8>>::new()));
+
+    let tx_infos = stream::iter(&block.transactions)
+        .then(|tx_hash| {
+            let accounts_state = accounts_state.clone();
+            let provider = Arc::clone(&provider);
+            let code_db = Arc::clone(&code_db);
+            async move {
+                super::txn::process_transaction(provider, tx_hash, accounts_state, code_db).await
+            }
+        })
```

Comment:
q: does the order in which we update the accounts state matter? if not we could use `FuturesOrdered` instead to execute the processing of transaction concurrenlt for a better performance. 

if the order matter, I'm curious why can't we `fetch_tx_data` here instead and update the accounts_state synchronously without using arc/lock as we have to process the txs sequentially

Reply:
I've got a change locally which changes this logic. I'll push them shortly.